tsl/src/nodes/decompress_chunk/planner.c

```diff
@@ -488,69 +501,129 @@ decompress_chunk_plan_create(PlannerInfo *root, RelOptInfo *rel, CustomPath *pat
 	/* Build heap sort info for sorted_merge_append */
 	List *sort_options = NIL;
 
-	if (dcpath->sorted_merge_append)
+	if (dcpath->batch_sorted_merge)
 	{
-		/* sorted_merge_append is used when the 'order by' of the query and the
-		 * 'order by' of the segments do match, we use a heap to merge the segments.
-		 * For the heap we need a compare function that determines the heap order. This
-		 * function is constructed here.
-		 */
-		AttrNumber *sortColIdx = NULL;
-		Oid *sortOperators = NULL;
-		Oid *collations = NULL;
-		bool *nullsFirst = NULL;
-		int numsortkeys = 0;
-
-		/* We need a targetlist at this point to build the sort info below. If the target list is
-		 * not populated by PostgreSQL already, populate it here.
+		/*
+		 * 'order by' of the query and the 'order by' of the compressed  batches
+		 * match, so we will we use a heap to merge the batches. For the heap we
+		 * need a compare function that determines the heap order. This function
+		 * is constructed here.
+		 *
+		 * Batch sorted merge is done over the decompressed chunk scan tuple, so
+		 * we must match the pathkeys to the decompressed chunk tupdesc.
 		 */
-		if (decompress_plan->scan.plan.targetlist == NIL)
-			decompress_plan->scan.plan.targetlist = ts_build_path_tlist(root, (Path *) path);
-
-		Assert(decompress_plan->scan.plan.targetlist != NIL);
-		Assert(dcpath->cpath.path.pathkeys != NIL);
-
-		ts_prepare_sort_from_pathkeys(&decompress_plan->scan.plan,
-									  dcpath->cpath.path.pathkeys,
-									  bms_make_singleton(dcpath->info->chunk_rel->relid),
-									  NULL,
-									  false,
-									  &numsortkeys,
-									  &sortColIdx,
-									  &sortOperators,
-									  &collations,
-									  &nullsFirst);
+
+		int numsortkeys = list_length(dcpath->custom_path.path.pathkeys);
 
 		List *sort_col_idx = NIL;
 		List *sort_ops = NIL;
 		List *sort_collations = NIL;
 		List *sort_nulls = NIL;
 
-		/* Since we have to keep the sort info in custom_private, we store the information
-		 * in copyable lists */
-		for (int i = 0; i < numsortkeys; i++)
+		/*
+		 */
+		ListCell *lc;
+		foreach (lc, dcpath->custom_path.path.pathkeys)
 		{
-			sort_col_idx = lappend_oid(sort_col_idx, sortColIdx[i]);
-			sort_ops = lappend_oid(sort_ops, sortOperators[i]);
-			sort_collations = lappend_oid(sort_collations, collations[i]);
-			sort_nulls = lappend_oid(sort_nulls, nullsFirst[i]);
+			PathKey *pk = lfirst(lc);
+			EquivalenceClass *ec = pk->pk_eclass;
+
+			/*
+			 * Find the equivalence member that belongs to decompressed relation.
+			 */
+			ListCell *membercell = NULL;
+			foreach (membercell, ec->ec_members)
+			{
+				EquivalenceMember *em = lfirst(membercell);
+
+				if (em->em_is_const)
+				{
+					continue;
+				}
+
+				Index em_relid;
```

Comment:
Shouldn't this be an int?

Reply:
Yes, let's change, although I think the comparison below will require a cast to Index anyway.